### ISSUE TITLE
NodeRule

### ISSUE THREAD

**Issue by synkarius** [synkarius/caster#53]
- status window hinting children node names
- Continuer + node for reset
- do sample node: CSS
- integrate with CCR

**Comment by synkarius** [comment]
status window hinting works, CSS node mostly works

**Comment by synkarius** [comment]
continuer works

**Comment by synkarius** [comment]
everything works

**Comment by Zone22** [comment]
I can't figure it out to register a NodeRule. If only one Node Rule is present everything works as expected so I'm doing something wrong with the syntax?
<pre><code>
_mapping={}
for node in [# register nodes here in order to get them into ccr
                css.getCSSNode()
                python.getPYTHONNode()
                 ]:
</pre></code>

**Comment by synkarius** [comment]
You're close. You need a comma between them.

**Comment by synkarius** [comment]
Also, if you're doing a Python NodeRule, there are some things you should be aware of.

* at present, NodeRule is hardcoded such that you can only speak two nodes at a time -- I can and probably should make that hardcoded value a parameter, but the more nodes you're allowed to speak in a single utterance, the less useful NodeRule is in general, because it prohibits less from being in the active grammar
* unlike normal CCR, NodeRule commands cause a slight pause after being executed because they have to make adjustments to their own grammar to activate the next node(s) or reset, which involves compatibility checking with everything else in active CCR -- it's possible that the compatibility checking could be optimized, but I have no ideas for this at present

In other words, there are trade-offs to doing a command set as a NodeRule. The major benefit is that for command sets which are treelike (the vast majority being commands which will only ever be spoken after other commands), NodeRule is a natural fit and allows for virtually limitless amounts of commands to be nested. 

Personally, I wouldn't think of Python or most programming languages as being treelike. I think of them as being more of a flattened structure. However, I suppose you could create a command set that way, and I'm not against it if you want to try it out.

**Comment by Zone22** [comment]
At this point I'm just experimenting trying to get a basic NodeRule implemented. I'm going to have to change the name Python anyway due to compatibility.

I'm showing my ignorance here but could you post the correct syntax I've tried six different ways with a comma and not had success.

**Comment by synkarius** [comment]
Ah, I see. Sorry to jump to conclusions there. I just didn't want you to do a lot of work and then be surprised by the trade-offs if you weren't aware of them already. Anyway, you can just call the top level node "python node" (or similar) instead of "python" and that should work.

Happy to do an example of the reworked syntax. 

    H = HintNode #semantic sugar I used across the css node
    H("else if", Key("e,l,i,f"), [], [], {})

The last 3 params (children, extras, defaults) are optional, so the above node could be alternatively written as

    H("else if", Key("e,l,i,f"))

for the same effect.

I recently adjusted the NodeRule docs to reflect the new syntax, and there are now a bunch more examples there. I know sometimes my docs are a bit terse though, so if you need more, don't hesitate to ask.

**Comment by synkarius** [comment]
Oh, one other thing to be aware of, don't put extra commas in the children list. That caused me some grief before. Commas only between children, not after the last one in the list.
